emma-spark/src/main/scala/org/emmalanguage/api/SparkDataset.scala

```diff
@@ -214,4 +214,15 @@ object SparkDataset {
       case xs: SparkDataset[A] => xs.rep.cache()
       case _ => xs
     }
+
+  def foldGroup[A: Meta, B: Meta, K: Meta](
+    xs: DataBag[A], key: A => K, sng: A => B, uni: (B, B) => B
+  )(implicit spark: SparkSession): DataBag[(K, B)] = xs match {
+    case xs: SparkRDD[A] => spark.createDataset(xs.rep
+      .map(x => key(x) -> sng(x)).reduceByKey(uni))
+    case xs: SparkDataset[A] => xs.rep
+      .map(x => key(x) -> sng(x)).groupByKey(_._1)
+      .reduceGroups((x, y) => x._1 -> uni(x._2, y._2))
+      .map((kv: (K, (K, B))) => kv._2)
+  }
```

Comment:
Implement this case by going through an `rdd` conversion for the moment (this will avoid group materialization as opposed to the `reduceGroups` variant).

```scala
case xs: SparkDataset[A] => xs.rep.rdd
  .map(x => key(x) -> sng(x)).reduceByKey(uni)).toDS
```